spec/models/ems_refresh/refreshers/azure_refresher_spec.rb

```diff
@@ -0,0 +1,199 @@
+require "spec_helper"
+
+describe EmsAzure do
+  before(:each) do
+    _guid, _server, zone = EvmSpecHelper.create_guid_miq_server_zone
+    @ems = FactoryGirl.create(:ems_azure, :zone => zone)
+    @ems.update_authentication(:default => {
+      :userid => "f895b5ef-3bb5-4366-8ce5-123456789012", :password => "0%2BdVmFpJLU0T9gqFWDkWnX9MY3FPb5l1yY123456789012"})
+    @ems.update_attributes(:tenant_id => "a50f9983-d1a2-4a8d-be7d-123456789012")
+  end
+
+  it "will perform a full refresh" do
+    2.times do  # Run twice to verify that a second run with existing data does not change anything
+      @ems.reload
+
+      VCR.use_cassette(described_class.name.underscore) do
+        EmsRefresh.refresh(@ems)
+      end
+      @ems.reload
+
+      assert_table_counts
+      assert_ems
+      assert_specific_az
+      assert_specific_flavor
+      assert_specific_vm_powered_on
+      assert_specific_vm_powered_off
+    end
+  end
+
+  def assert_table_counts
+    ExtManagementSystem.count.should == 1
+    Flavor.count.should              == 41
+    AvailabilityZone.count.should    == 2
+    VmOrTemplate.count.should        == 5
+    Vm.count.should                  == 5
+    Disk.count.should                == 7
+    GuestDevice.count.should         == 0
+    Hardware.count.should            == 5
+    Network.count.should             == 10
+    OperatingSystem.count.should     == 5
+    Relationship.count.should        == 0
+    MiqQueue.count.should            == 7
+  end
+
+  def assert_ems
+    @ems.should have_attributes(
+      :api_version => nil,
+      :uid_ems     => "a50f9983-d1a2-4a8d-be7d-123456789012"
+    )
+    @ems.flavors.size.should              == 41
+    @ems.availability_zones.size.should   == 2
+    @ems.vms_and_templates.size.should    == 5
+    @ems.vms.size.should                  == 5
+  end
+
+  def assert_specific_flavor
+    @flavor = EmsAzure::Flavor.where(:name => "Standard_A1").first
+    @flavor.should have_attributes(
+      :name                     => "Standard_A1",
+      :description              => nil,
+      :enabled                  => true,
+      :cpus                     => 1,
+      :cpu_cores                => 1,
+      :memory                   => 1792,
+      :supports_32_bit          => nil,
+      :supports_64_bit          => nil,
+      :supports_hvm             => nil,
+      :supports_paravirtual     => nil,
+      :block_storage_based_only => nil,
+    )
+
+    @flavor.ext_management_system.should == @ems
+  end
+
+  def assert_specific_az
+    @az = EmsAzure::AvailabilityZone.where(:name => "AvailabilitySet1").first
```

Comment:
Is this only partially namespaced?  This is so confusing in this form.  cc @matthewd